### PR TITLE
Angular rule

### PR DESCRIPTION
**Pull request by remcohaszing**
I'm writing a rule wrapper which makes it easier to write some AngularJS related rules.

The goal is that common rule definitions don't have to implement the same common checks and lookups individually. Also this should help fixing common false positives.

Ultimately I think it is possible to achieve a wrapper which can resolve variables using ESLint scopes and understands the following are AngularJS components:
```js
var myApp = angular.module('myApp').factory('toyFactory', function() {});

myApp.filter('parentalControl', parentalControlFilter);

function parentalControlFilter() {}
```

Whereas these are not:
```js
_.filter(obj, coffeeFilter);

mocha.run();
```
Currently the component extraction is handled by each rule separately, which grealy impacts performance. For this reason this PR should not be merged yet.

I think this can be shared eventually. This way it might even become more performant than not using this rule, because many checks can be skipped in the rule definitions themselves.

For demo purposes I have ported `di-order` and `di-unused` to use this new setup.

@Gillespie59 @tilmanpotthof do you guys think this is a good approach?

### CODE REVIEW

File: test/di-order.js [hunk 1]
```diff
@@ -15,19 +15,17 @@ var commonFalsePositives = require('./utils/commonFalsePositives');
 var eslintTester = new RuleTester();
 eslintTester.run('di-order', rule, {
     valid: [
-        'app.controller("", function($http, $q) {});',
-        'app.directive("", function($http, $q) {});',
```
Comment (tilmanpotthof) [inline test/di-order.js]: Will the 'app.*' Syntax work as well? Although there is a rule to check module getters, this syntax should not be enforced by another rule.

**Comment by tilmanpotthof** [comment]
I like the approach to unify common things and have a abstraction for repeating angular specific logic. I just scanned the code but will have a deeper look at it within the next days.

File: test/di-order.js (hunk 1 above)
Comment (remcohaszing) [inline test/di-order.js]: ESLint tries to resolve identifiers. I would like to use this to make sure that the following is checked:
```
var app = angular.module('');
app.controller('', function() {});
```

But the following is ignored:
```
var app = someUnknownCall();
app.controller('', function() {})
```

**Comment by Gillespie59** [comment]
Looks very useful !!! will also have a look deeper to your implem ! But I Think this approach is mandatory, in order to avoid duplicate checks.

**Comment by remcohaszing** [comment]
I was concerned about the performance, so I tested it against an actual code base. It barely made a difference.

If you both agree with this setup, I think it's ready to be merged.

**Comment by Gillespie59** [comment]
Ok for me. 
What is the process to use in order to convert the rule with your wrapper ?

**Comment by remcohaszing** [comment]
Typically a `CallExpression` node is used in combination with `utils.isAngularComponent` to determine if a the node defines an Angular component. This is now changed to `angular:componentType` which gets called with the call expression and the resolved function expression / declaration.

```
return {
  'angular:service': function(callExpressionNode, fnNode) {}
};
```

Also tests need to be modified. E.g.
```
app.service("", function() {});
```
needs to become:
```
angular.module("").service("", function() {});
```
Also detected are:
```
var app = angular.module(""); app.service("", function() {});
```
And:
```
angular.module("").service("", myService);
function myService() {}
```

Currently there is only one test verifying the angular module variable resolving and the component function declaration resolving work.

Since ESLint scopes are used to determine variables, this can't be unittested using Espree (as far as I know).

Do you think every rule should test this individually?

Also I did not consider `animation` and `component` as AngularJS components when writing this rule. The rules changed in this PR didn't check these before either, but they should.

**Comment by palortoff** [comment]
Ah.  That's fine of course.

**Comment by remcohaszing** [comment]
@Gillespie59 told me I could merge this in #277.

**Comment by maksimr** [comment]
@remcohaszing this PR contains fix for #277?

**Pull request merged**